src/python/pants/backend/python/rules/coverage.py

```diff
@@ -0,0 +1,351 @@
+
+
+DEFAULT_COVERAGE_CONFIG = dedent(f"""
+  [run]
+  branch = True
+  timid = False
+  relative_files = True
+  """)
+
+
+def get_coverage_plugin_input():
+  return InputFilesContent(
+    FilesContent(
+      (
+        FileContent(
+          path=f'{COVERAGE_PLUGIN_MODULE_NAME}.py',
+          content=pkg_resources.resource_string(__name__, 'coverage_plugin/plugin.py'),
+        ),
+      )
+    )
+  )
+
+
+def get_coveragerc_input(coveragerc_content: str) -> InputFilesContent:
+  return InputFilesContent(
+    [
+      FileContent(
+        path='.coveragerc',
+        content=coveragerc_content.encode(),
+        is_executable=False,
+      ),
+    ]
+  )
+
+
+COVERAGE_PLUGIN_MODULE_NAME = '__coverage_coverage_plugin__'
+
+
+def ensure_section(config_parser: configparser.ConfigParser, section: str) -> None:
+  """Ensure a section exists in a ConfigParser."""
+  if not config_parser.has_section(section):
+    config_parser.add_section(section)
+
+
+def construct_coverage_config(
+  source_roots: SourceRoots,
+  python_files: List[str],
+  test_time: Optional[bool] = False,
+) -> str:
+  # A map from source root stripped source to its source root. eg:
+  #  {'pants/testutil/subsystem/util.py': 'src/python'}
+  # This is so coverage reports referencing /chroot/path/pants/testutil/subsystem/util.py can be mapped
+  # back to the actual sources they reference when merging coverage reports.
+  init_files = list(identify_missing_init_files(list(python_files)))
+
+  def source_root_stripped_source_and_source_root(file_name):
+    source_root = source_roots.find_by_path(file_name)
+    source_root_stripped_path = file_name[len(source_root.path)+1:]
+    return (source_root_stripped_path, source_root.path)
+
+  source_to_target_base = dict(
+    source_root_stripped_source_and_source_root(filename) for filename in python_files + init_files
+  )
+  config_parser = configparser.ConfigParser()
+  config_parser.read_file(StringIO(DEFAULT_COVERAGE_CONFIG))
+  ensure_section(config_parser, 'run')
+  config_parser.set('run', 'plugins', COVERAGE_PLUGIN_MODULE_NAME)
+  config_parser.add_section(COVERAGE_PLUGIN_MODULE_NAME)
+  config_parser.set(COVERAGE_PLUGIN_MODULE_NAME, 'source_to_target_base', json.dumps(source_to_target_base))
+  config_parser.set(COVERAGE_PLUGIN_MODULE_NAME, 'test_time', json.dumps(test_time))
+  config = StringIO()
+  config_parser.write(config)
+  return config.getvalue()
+
+
+class ReportType(Enum):
+  XML = "xml"
+  HTML = "html"
+
+
+class CoverageOptions(GoalSubsystem):
+  """Runs pytest coverage."""
+  name = 'coverage2'
+
+
+class PytestCoverage(PythonToolBase):
+  options_scope = 'pytest-coverage'
+  default_version = 'coverage==5.0.3'
+  default_entry_point = 'coverage'
+  default_interpreter_constraints = ["CPython>=3.6"]
+
+  @classmethod
+  def register_options(cls, register):
+    super().register_options(register)
+    register(
+      '--report-output-path',
+      type=str,
+      default='dist/coverage/python',
```

Comment:
This should probably be `os.path.join(DistDir.relpath, "coverage", "python")` rather than hard-coding `dist`

Reply:
It's not clear to me I can fetch this from the engine in the `PythonToolBase` where I'm registering options. 

If I use DistDir here I think I have to construct it myself and then it looks like this:
`os.path.join(DistDir(relpath='dist').relpath, 'coverage', 'python')`

which doesn't seem like much of an improvement. Am I missing a better way to get an instance of DistDir outside of a rule?